test/core/linted_file_test.py

```diff
@@ -0,0 +1,350 @@
+"""The Test file for the linter class."""
+
+import pytest
+import logging
+
+from sqlfluff.core.linter import LintedFile
+from sqlfluff.core.parser.markers import PositionMarker
+from sqlfluff.core.parser.segments import (
+    FixPatch,
+    RawSegment,
+    BaseSegment,
+    TemplateSegment,
+)
+from sqlfluff.core.parser.segments.raw import SourceFix
+from sqlfluff.core.templaters import RawFileSlice, TemplatedFile
+from sqlfluff.core.templaters.base import TemplatedFileSlice
+
+
+@pytest.mark.parametrize(
+    "source_slices,source_patches,raw_source_string,expected_result",
+    # NOTE: For all of these examples we're not setting the patch_category
+    # of the fix patches. They're not used at this step so irrelevant for
+    # testing.
+    [
+        # Trivial example
+        ([slice(0, 1)], [], "a", "a"),
+        # Simple replacement
+        (
+            [slice(0, 1), slice(1, 2), slice(2, 3)],
+            [FixPatch(slice(1, 2), "d", "", slice(1, 2), "b", "b")],
+            "abc",
+            "adc",
+        ),
+        # Simple insertion
+        (
+            [slice(0, 1), slice(1, 1), slice(1, 2)],
+            [FixPatch(slice(1, 1), "b", "", slice(1, 1), "", "")],
+            "ac",
+            "abc",
+        ),
+        # Simple deletion
+        (
+            [slice(0, 1), slice(1, 2), slice(2, 3)],
+            [FixPatch(slice(1, 2), "", "", slice(1, 2), "b", "b")],
+            "abc",
+            "ac",
+        ),
+        # Illustrative templated example (although practically at
+        # this step, the routine shouldn't care if it's templated).
+        (
+            [slice(0, 2), slice(2, 7), slice(7, 9)],
+            [FixPatch(slice(2, 3), "{{ b }}", "", slice(2, 7), "b", "{{b}}")],
+            "a {{b}} c",
+            "a {{ b }} c",
+        ),
+    ],
+)
+def test__linted_file__build_up_fixed_source_string(
+    source_slices, source_patches, raw_source_string, expected_result, caplog
+):
+    """Test _build_up_fixed_source_string.
+
+    This is part of fix_string().
+    """
+    with caplog.at_level(logging.DEBUG, logger="sqlfluff.linter"):
+        result = LintedFile._build_up_fixed_source_string(
+            source_slices, source_patches, raw_source_string
+        )
+    assert result == expected_result
+
+
+@pytest.mark.parametrize(
+    "source_patches,source_only_slices,raw_source_string,expected_result",
+    # NOTE: For all of these examples we're not setting the patch_category
+    # of the fix patches. They're not used at this step so irrelevant for
+    # testing.
+    [
```

Comment:
I'm having trouble grasping the test cases for this test `test__linted_file__slice_source_file_using_patches` -- could you add more detailed comments explaining one or two of them in more detail?

Reply:
I've added some more detailed comments on all of them explaining what's supposed to happen. I think there's definitely some room for streamlining here however. Especially now that the tracing is pretty accurate I wonder whether some of the safety measures in here are a little unnecessary now.

I'm not planning on addressing that here, but I've left the TODOs in the code as a reminder for later than there are some things here that we might want to work on in future. I think I've been looking at this code for too long to see any more obvious quick wins 😬 .